R/MsExperiment.R

```diff
@@ -214,48 +214,99 @@ setMethod("readMsObject", signature(object = "MsExperiment",
+                   simplify = TRUE, ...) {
+              if (!requireNamespace("MsBackendMetaboLights", quietly = TRUE)) {
+                  stop("Required package 'MsBackendMetaboLights' is missing. ",
+                       "Please install it and try again.", call. = FALSE)
+              }
+              pth <- MsBackendMetaboLights::mtbls_ftp_path(param@mtblsId)
+              all_fls <- MsBackendMetaboLights::mtbls_list_files(param@mtblsId)
 
-              ## check assay files
-              assays <- grep("a_", files, value = TRUE)
-              if (length(assays) > 1) {
+              ## Extract and read assay files
+              assays <- all_fls[grepl("^a_", all_fls)]
+              if (length(param@assayName) > 0) selected_assay <- param@assayName
+              else if (length(assays) > 1 && length(param@assayName) == 0) {
                   cat("Multiple assay files found:\n")
                   selection <- menu(assays,
                                     title = paste("Please choose the assay",
                                     "file you want to use:"))
                   selected_assay <- assays[selection]
               } else if (length(assays) == 1) {
-                  selected_assay <- assays[1]
+                  selected_assay <- assays
                   cat("Only one assay file found:", selected_assay, "\n")
-              } else stop("No assay files found.") ## i don't think that would happen...
-
-              assay_data <- read.table(file.path(url, selected_assay),
-                                       header = TRUE, sep = "\t")
-              assay_data$injection_index <- seq_len(nrow(assay_data))
+              }
+              assay_data <- read.table(paste0(pth, selected_assay),
+                                       header = TRUE, sep = "\t",
+                                       check.names = FALSE)
 
               ## Extract and read sample info files
-              sample_info_files <- grep("s_", files, value = TRUE)
-              sample_info <- read.table(file.path(url, sample_info_files),
-                                        header = TRUE, sep = "\t")
-              merged_data <- merge(assay_data, sample_info, by = "Sample.Name")
-              merged_data <- merged_data[order(merged_data$injection_index), ]
-              merged_data <- merged_data[, colSums(is.na(merged_data)) <
-                                             nrow(merged_data)]
+              s_files <- all_fls[grepl("^s_", all_fls)]
+              sample_info <- read.table(paste0(pth, s_files),
+                                        header = TRUE, sep = "\t",
+                                        check.names = FALSE)
+
+              # merging
+              ord <- match(assay_data$`Sample Name`, sample_info$`Sample Name`)
+              merged_data <- cbind(assay_data, sample_info[ord, ])
+              names(merged_data) <- gsub(" ", "_", names(merged_data))
```

Comment:
I would not rename the column names (see comment below for `object@sampleData <- DataFrame(...`)